### PR TITLE
Tensorflow integration

### PR DESCRIPTION
**Pull request by kevinmartinjos**
#### New Rerankers:

- TFKNRM
- TFVanillaBERT

#### New Trainer:
- TensorflowTrainer

#### To train using TPUs:

```
python capreolus/run.py rerank.train with reranker=TFKNRM benchmark=wsdm20demo collection=robust04 reranker.trainer.niters=20 reranker.trainer.usecache=True reranker.trainer.tpuname=node-6 reranker.trainer.tpuzone=us-central1-a reranker.trainer.storage=gs://bucketname/somedir searcher=bm25staticrob04yang19
```

#### Misc

- Only the v3 tpu named `node-6` works for the time being - other TPUs have some firmware mismatch with tensorflow 2.1
- Authentication is handled manually. I made my gcs bucket public. Alternatively, I was able to do the auth by attempting to run the training, noting down the tpu address from the resulting 403 error, and then granting this tpu read/write rights through the cloud storage web page.  

#### TFKNRM Accuracy
Around 0.36 ndcg_20 on robust04 with gradkernels=False. Training on TPU is slower than CPU (!!) due to some input constraints. 

#### TFVanillaBERT Accuracy

Still waiting for a complete run (niters=50) to finish. Got a dev ndcg_20 of 0.424 (at niter=3). The validation metrics decrease as niters increases (overfitting?)

#### To Fix:
Github actions - with tensorflow being added to requirements.txt, the github actions errors out due to insufficient space. The unit tests pass on local though

### CODE REVIEW

File: capreolus/trainer/__init__.py [hunk 1]
```diff
@@ -292,3 +297,240 @@ def predict(self, reranker, pred_data, pred_fn):
...
+        lr = 0.001  # learning rate
+        softmaxloss = False  # True to use softmax loss (over pairs) or False to use hinge loss
+
+        interactive = False  # True for training with Notebook or False for command line environment
+        fastforward = False
+        validatefreq = 1
+        usecache = False
+
+    def get_optimizer(self):
+        return tf.keras.optimizers.Adam(learning_rate=self.cfg['lr'])
```
Comment (kevinmartinjos) [inline capreolus/trainer/__init__.py:345]: TODO: Make this based on a passed config

File: capreolus/trainer/__init__.py [hunk 2]
```diff
@@ -292,3 +297,240 @@ def predict(self, reranker, pred_data, pred_fn):
...
+        for niter in tqdm(range(0, self.cfg["niters"]), desc="Converting data to tf records"):
+            for sample_idx, data in enumerate(dataset):
+                # TODO: Split into multiple files. This will destroy the RAM
+                tf_features.append(
+                    self.create_tf_feature(
+                        data["qid"], data["query"], data["query_idf"], data["posdocid"], data["posdoc"],
+                        data["negdocid"], data["negdoc"]
+                    )
+                )
+                if sample_idx + 1 >= self.cfg["itersize"] * self.cfg["batch"]:
```
Comment (kevinmartinjos) [inline capreolus/trainer/__init__.py:476]: I probable need a better strategy here. What I do now:

1. Iterate through the dataset to get `niters * itersize * batch` number of samples
2. Save the ^ above samples to a tf record
3. When I make use of the tf record, I can do a simple iteration (i.e `for x in tf_record`) without bothering about itersize or niters since the tf records already contains the exact number of samples I need.
4. This might pose problems for caching. Eg: I create tf records and cache it for niters=10, itersize=32, batch=16. Next time I train the tf records will be read from the cache and if I set my batch size as 32 this time, it would mean that I would not have enough training samples in my tf record (assuming I'd still want niters=10 and itersize=32).

Comment (andrewyates) [reply capreolus/trainer/__init__.py:518]: Two possible solutions:

1. We could make the cache key a function of the sampling parameters and the total number of samples `total_samples = niters * itersize * batch`. Then when looking for a cached tfrecord, we load any that has <= the number of samples we need.

2. We could write the unique samples out once and then use [tf.data.Dataset.shuffle](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle) to sample from them. This is what Crystina's TFv1 implementation does. If we go for this option, we need to be careful to make sure `shuffle`'s `buffer_size` is large enough that we see new data every iteration. When writing the unique samples, we still need to be careful to iterate over them in the right order (e.g., round robin over queries).

The second option is more straightforward but reduces our control over sampling. I think this is okay though.

File: capreolus/trainer/__init__.py [hunk 3]
```diff
@@ -292,3 +322,265 @@ def predict(self, reranker, pred_data, pred_fn):
...
+    def load_tf_records_from_file(self, filenames):
+        raw_dataset = tf.data.TFRecordDataset(filenames)
+        feature_description = {
+            'qid': tf.io.FixedLenFeature([], tf.string),
+            'query': tf.io.FixedLenFeature([self.cfg["maxqlen"]], tf.float32),
+            'query_idf': tf.io.FixedLenFeature([self.cfg["maxqlen"]], tf.float32),
+            'posdoc_id': tf.io.FixedLenFeature([], tf.string),
+            'posdoc': tf.io.FixedLenFeature([self.cfg["maxdoclen"]], tf.float32),
+            'negdoc_id': tf.io.FixedLenFeature([], tf.string, default_value=b'na'),
+            'negdoc': tf.io.FixedLenFeature([self.cfg['maxdoclen']], tf.float32, default_value=tf.zeros(self.cfg['maxdoclen']))
```
Comment (andrewyates) [inline capreolus/trainer/__init__.py:548]: Will the cache be broken if `maxqlen` or `maxdoclen` changes?

File: capreolus/trainer/__init__.py (hunk 2 above)
Comment (kevinmartinjos) [reply capreolus/trainer/__init__.py:518]: Hmmmm so for option 2, you are suggesting that we think in terms of epochs right? An epoch in this case would be `num_queries * num_posdocs_in_qrels * num_negdocs_in_qrels`. So all we have to do is save one epoch and then keep shuffling it every time we load it? This sounds like the better approach

File: capreolus/trainer/__init__.py (hunk 3 above)
Comment (kevinmartinjos) [reply capreolus/trainer/__init__.py:548]: Hmmm yes it would be. We iterate through the dataset, convert everything using the `extractor` and then cache the results as tf records. If the maxqlen changes, or if the embedding (glove6b) changes, then the cache would need to be invalidated.

I'll add the extractor config as well contribute to the dataset's `__hash()`

File: capreolus/trainer/__init__.py [hunk 4]
```diff
@@ -292,3 +322,265 @@ def predict(self, reranker, pred_data, pred_fn):
...
+
+        if negdoc_id:
+            feature["negdoc_id"] = tf.train.Feature(bytes_list=tf.train.BytesList(value=[negdoc_id.encode('utf-8')])),
+            feature["negdoc"] = tf.train.Feature(float_list=tf.train.FloatList(value=negdoc))
+
+        return feature
+
+    def write_tf_record_to_file(self, dir_name, tf_features):
+        filename = dir_name / "{}.tfrecord".format(str(uuid.uuid4()))
+        examples = [tf.train.Example(features=tf.train.Features(feature=feature)) for feature in tf_features]
```
[the remark this answers is not in the document]
Comment (kevinmartinjos) [reply capreolus/trainer/__init__.py:471]: We might have to ditch generators altogether - if we use generators here, a "PyFunc" gets included in the resulting tf graph and it thus can't be executed on TPUs because a TPU does not have a python interpreter
